### ISSUE TITLE
LexicalPreservingPrinter fails with Could not parse error

### ISSUE THREAD

**Issue by rohitsureka** [javaparser/javaparser#2293]
Version: 3.14.5
I'm running into an issue where I'm trying to use the LexicalPreservingPrinter after altering the AST and it throws an exception
```
Exception in thread "main" java.lang.Error: Could not parse : java.lang.IllegalArgumentException: fromIndex(14) > toIndex(10)
	at java.util.SubList.<init>(AbstractList.java:624)
	at java.util.AbstractList.subList(AbstractList.java:484)
	at com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator$CalculatedSyntaxModel.sub(LexicalDifferenceCalculator.java:44)
	at com.github.javaparser.printer.lexicalpreservation.DifferenceElementCalculator.calculate(DifferenceElementCalculator.java:123)
	at com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator.calculateListReplacementDifference(LexicalDifferenceCalculator.java:110)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter$Observer.concreteListReplacement(LexicalPreservingPrinter.java:280)
	at com.github.javaparser.ast.observer.PropagatingAstObserver.listReplacement(PropagatingAstObserver.java:82)
	at com.github.javaparser.ast.NodeList.lambda$notifyElementReplaced$3(NodeList.java:474)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.github.javaparser.ast.NodeList.notifyElementReplaced(NodeList.java:474)
	at com.github.javaparser.ast.NodeList.set(NodeList.java:147)
	at com.github.javaparser.ast.NodeList.set(NodeList.java:47)
	at com.github.javaparser.utils.Utils.replaceElementByObjectIdentity(Utils.java:232)
	at com.github.javaparser.ast.visitor.ModifierVisitor.visit(ModifierVisitor.java:1077)
```
It works fine when I try to not use the LexicalPreservingPrinter and simply do
```
compilationUnitObject.toString()
```
ofcourse in that case, it doesn't preserve my formatting. Has anyone run into an exception with a similar stack trace / use case. I can help provide a minimalistic test case for the same if thats helpful

**Comment by edefazio** [comment]
Yes, I've had that happen to me, and it seemed to happen consistently when i made "batch changes"
(i.e. not a simple change like adding a modifier, etc., but adding multiple lines of code in different sections of the code)

I eventually had to give up on using the LPP... but (on the plus side)
Have you looked at the output and seen the regular formatting? 
(w/o the LexicalPreservingPrinter) and seen it? 

Much as I thought it'd bother me, it actually retained most of the formatting

(In my case the functionality >>>> code formatting)

...And you always have the option of writing your own JavaParser Code formatter... (it's pretty easy IMHO)

**Comment by rohitsureka** [comment]
@edefazio thanks for your input, I might consider doing that. I am actually using the LPP for basically just "preserving the empty lines" within my code. For everything else, I'm can use the PrettyPrinterConfiguration. 
Can you point me to an example of writing your own JavaParser code formatter if you have it handy ?

**Comment by edefazio** [comment]
Here's what I did... this is sorta easy, but it's best if you play around with the 
input to the `PrettyPrintVisitor`

````java
  /**
     * An instance of a {@link com.github.javaparser.ast.visitor.VoidVisitor}
     * used to print out entities, will specifically Avoid printing ANNOTATIONS
     * by simply overriding specific annotation METHODS with no-ops:
     *
     * @see #PRINT_NO_ANNOTATIONS_OR_COMMENTS
     *
     * This is also an example of an implementation of a Visitor for generating
     * the formatted .java source code, for the more comprehensive example:
     * @see com.github.javaparser.printer.PrettyPrintVisitor
     */
    public static class PrintNoAnnotations extends PrettyPrintVisitor {

        public PrintNoAnnotations(PrettyPrinterConfiguration prettyPrinterConfiguration) {
            super(prettyPrinterConfiguration);
        }

        private void printMemberAnnotations(NodeList<AnnotationExpr> annotations, Void arg) {
        }

        private void printAnnotations(NodeList<AnnotationExpr> annotations, boolean prefixWithASpace, Void arg) {
        }

        @Override
        public void visit(MarkerAnnotationExpr n, Void arg) {
        }

        @Override
        public void visit(SingleMemberAnnotationExpr n, Void arg) {
        }

        @Override
        public void visit(NormalAnnotationExpr n, Void arg) {
        }
    }

    public static final PrettyPrinterConfiguration PRINT_NO_COMMENTS = new PrettyPrinterConfiguration()
            .setPrintComments(false).setPrintJavadoc(false);

    public static class PrintRawComments extends PrettyPrintVisitor {

        public PrintRawComments(PrettyPrinterConfiguration prettyPrinterConfiguration) {
            super(prettyPrinterConfiguration);
        }

        @Override
        public void visit(final LineComment n, final Void arg) {
            if (configuration.isIgnoreComments()) {
                return;
            }
            printer.println(normalizeEolInTextBlock(n.getContent(), "").trim());
        }

        @Override
        public void visit(final BlockComment n, final Void arg) {
            if (configuration.isIgnoreComments()) {
                return;
            }
            String[] lines = n.getContent().split("\\r?\\n");
            StringBuilder sb = new StringBuilder();
            Arrays.stream(lines).forEach(con -> {
                String tr = con.trim();
                if (tr.startsWith("* ")) {
                    tr = tr.substring(2).trim();
                    sb.append(tr).append(System.lineSeparator());
                } else {
                    sb.append(con.trim()).append(System.lineSeparator());
                }
            });
            String con = sb.toString().trim();
            printer.print(con);
        }

        @Override
        public void visit(final JavadocComment n, final Void arg) {
            if (configuration.isPrintComments() && configuration.isPrintJavadoc()) {
                String[] lines = n.getContent().split("\\r?\\n");

                StringBuilder sb = new StringBuilder();
                Arrays.stream(lines).forEach(con -> {
                    String tr = con.trim();
                    if (tr.startsWith("* ")) {
                        tr = tr.substring(2).trim();
                        sb.append(tr).append(System.lineSeparator());
                    } else {
                        sb.append(tr).append(System.lineSeparator());
                    }
                });
                String con = sb.toString().trim(); //n.getContent();
                printer.print(con);
            }
        }
    }

    public static final PrettyPrinterConfiguration PRINT_RAW_COMMENTS = new PrettyPrinterConfiguration()
            .setVisitorFactory(PrintRawComments::new);
    /**
     * a PrettyPrinterConfiguration designed to
     * <PRE>
     *  usage examples:
     *  _class _c = _class.of("C", new Object(){ @Deprecated int f; });
     *  _c.toString( Ast.PRINT_NO_ANNOTATIONS_OR_COMMENTS );
     * </PRE>
     */
    public static final PrettyPrinterConfiguration PRINT_NO_ANNOTATIONS_OR_COMMENTS
            = new PrettyPrinterConfiguration()
                    .setPrintComments(false).setPrintJavadoc(false).setVisitorFactory(PrintNoAnnotations::new);
````

**Comment by edefazio** [comment]
If you look at the source of `PrettyPrintVisitor` it's really not complicated... I tweaked it to do some different things and it's "shallow" and easily testable

probably want to override it and just extend/change what you want to change will be the easiest route

**Comment by rohitsureka** [comment]
@edefazio thanks, I just checked out your example. I pretty much ended up doing the same thing. I ended up calculating the the difference in between the Range of the two nodes and adding those empty lines! Thanks for your prompt response and help! Appreciate it.

**Comment by matozoid** [comment]
Cool solution!
Can this be closed? The original problem is the same as #2290.

**Comment by rohitsureka** [comment]
yes @matozoid, we can close this! Seems like changing the node type in the orignal AST causes the LexicalPreservingPrinter to barf is covered by #2290
